libs/s25main/ingameWindows/iwMilitaryBuilding.cpp

```diff
@@ -204,11 +243,48 @@ void iwMilitaryBuilding::Msg_ButtonClick(const unsigned ctrl_id)
             }
         }
         break;
-        case 10: // send home button (addon)
+        case 10:
         {
-            gcFactory.SendSoldiersHome(building->GetPos());
+            if(gwv.GetWorld().GetGGS().getSelection(AddonId::MILITARY_CONTROL) == 1)
+            {
+                // Send the highest rank soldiers in this building home and get new soldiers
+                if(building->GetNumTroops() > 1)
+                {
+                    auto maxRank = building->GetTroops().back().GetRank();
+                    gcFactory.SetTroopLimit(building->GetPos(), maxRank, 0);
+                    gcFactory.SetTroopLimit(building->GetPos(), maxRank, building->GetMaxTroopsCt());
+                }
+            }
+            else
+            {
+                // Send all soldiers home except one
+                gcFactory.SetTroopLimit(building->GetPos(), 0, 1);
+                for(unsigned rank = 1; rank < NUM_SOLDIER_RANKS; ++rank)
+                    gcFactory.SetTroopLimit(building->GetPos(), rank, 0);
+            }
         }
         break;
+        default:
+        {
+            if(ctrl_id > 10)
+            {
+                if(ctrl_id % NUM_SOLDIER_RANKS == 4)
```

Comment:
This isn't correct. The IDs for increasing the rank are `11 + (5 * i)`, i.e. not related to `NUM_SOLDIER_RANKS` although that is `5` (currently) by coincidence.

So what we have:
- MINUS button: `11 + (5 * rank) = 10 + 1 + (5 * rank)`
- PLUS button `14 + (5 * rank) = 10 + 4 + (5 * rank)`

So the `ctrl_id % x` is the same `x` as in `rank = (ctrl_id - y) / x`, i.e. 5. The modulo should just strip off the `(5 * rank)` and coincidently the 10 leaving only the 1 and 4.
Hence:
```suggestion
                if(ctrl_id % 5 == 4)
```